File: server/scraper/app/services/subscore_eval.py
"""Golden-set sub-score evaluation for the Evaluator (`POST /api/match`).

Unlike verdict_eval (which checks the overall verdict against a single
expected band), this checks individual `breakdown.*.score` dimensions
against a frozen candidate profile — so a failing case means the prompt
changed, not that the live Mongo profile changed underneath the eval.

Each case names the dimension(s) it cares about in `expected`
(e.g. {"technicalFit": "high"}) and is scored via `score_job(..., profile=...)`
with the frozen profile loaded from tests/fixtures/golden-profile.json on
every call. Dimensions not named in a case's `expected` are read (and
reported, for reference) but never compared.

Fails loud, same as verdict_eval: if any case's API call errors, the whole
run raises rather than report a partial result.

Run: python -m app.cli eval-subscore
"""
import json
import asyncio
import logging
from pathlib import Path

from app.config import Settings
from app.services import match_client
from app.services.verdict_eval import load_golden_set

logger = logging.getLogger(__name__)

GOLDEN_PROFILE_PATH = Path(__file__).resolve().parents[2] / "tests" / "fixtures" / "golden-profile.json"

# Same bucket as verdict_eval — /api/match is rate-limited under "match" (10/min).
PACING_SECONDS = 3.0

DIMENSIONS = ["technicalFit", "engineeringExecutionFit", "sustainabilityPaceFit"]

# Band a sub-score by percentage of its own maxScore (not a hardcoded point
# value — maxScore is model-reported and, per PromptSeeds, dimension-specific).
BAND_LOW_CEILING = 28   # < 28%  -> low
BAND_MID_CEILING = 57   # 28-57% -> mid; > 57% -> high


def load_golden_profile() -> dict:
    return json.loads(GOLDEN_PROFILE_PATH.read_text(encoding="utf-8"))


def band_for(score: int | None, max_score: int | None) -> str | None:
    """None when the response didn't carry a usable score/maxScore pair —
    kept distinct from a real band so a missing dimension never silently
    reads as a pass or a specific band."""
    if score is None or not max_score:
        return None
    pct = (score / max_score) * 100
    if pct < BAND_LOW_CEILING:
        return "low"
    if pct <= BAND_MID_CEILING:
        return "mid"
    return "high"


async def run_eval(settings: Settings, cases: list[dict] | None = None) -> list[dict]:
    """Score every golden-set case sequentially (paced) against the frozen
    golden-profile.json and return one result dict per case. Raises
    RuntimeError immediately if any call fails — never returns a partial
    result (same contract as verdict_eval.run_eval)."""
    cases = cases if cases is not None else load_golden_set()
    profile = load_golden_profile()
    results = []
    for i, case in enumerate(cases):
        if i > 0:
            await asyncio.sleep(PACING_SECONDS)
        response = await match_client.score_job(settings, case["jdText"], profile=profile)
        if response is None:
            raise RuntimeError(
                f"eval-subscore: API call failed for case '{case['id']}' — "
                "aborting the whole run rather than reporting a partial result."
            )

        breakdown = response.get("breakdown") or {}
        observed = {}
        for dim in DIMENSIONS:
            d = breakdown.get(dim) or {}
            score, max_score = d.get("score"), d.get("maxScore")
            observed[dim] = {"score": score, "maxScore": max_score, "band": band_for(score, max_score)}

        hard_blockers = response.get("hardBlockers") or []
        expect_no_blockers = bool(case.get("expectNoBlockers", False))
        no_blockers_ok = not (expect_no_blockers and hard_blockers)

        # expectBlocked is the inverse, tri-state check: True/False actually
        # assert on hardBlockers being non-empty/empty; absent (None) means no
        # opinion from this flag (expectNoBlockers, above, still applies on
        # its own). Cases that test the veto itself carry no `expected` dict.
        expect_blocked = case.get("expectBlocked")
        if expect_blocked is None:
            blocked_ok = True
        else:
            blocked_ok = bool(hard_blockers) == bool(expect_blocked)

        blockers_ok = no_blockers_ok and blocked_ok

        expected = case.get("expected") or {}
        checked = []
        for dim, expected_band in expected.items():
            actual = observed.get(dim, {"score": None, "maxScore": None, "band": None})
            checked.append({
                "dimension": dim,
                "expectedBand": expected_band,
                "actualBand": actual["band"],
                "score": actual["score"],
                "maxScore": actual["maxScore"],
                "passed": actual["band"] == expected_band,
            })

        results.append({
            "id": case["id"],
            "title": case.get("title"),
            "probe": case.get("probe"),
            "signal": case.get("signal"),
            "blocker": case.get("blocker"),
            "dependsOnProfile": bool(case.get("dependsOnProfile", False)),
            "checked": checked,
            "expectNoBlockers": expect_no_blockers,
            "expectBlocked": expect_blocked,
            "hardBlockers": hard_blockers,
            "blockersOk": blockers_ok,
            "observed": observed,
            "passed": blockers_ok and all(c["passed"] for c in checked),
        })
    return results


def format_report(results: list[dict]) -> str:
    total = len(results)
    passed = sum(1 for r in results if r["passed"])

    lines = [
        f"Evaluator sub-score golden-set report — {passed}/{total} passed",
        "",
    ]

    for r in results:
        status = "PASS" if r["passed"] else "FAIL"
        header = f"[{status}] {r['id']}"
        if r.get("title"):
            header += f" — {r['title']}"
        meta = [f"{k}={r[k]}" for k in ("probe", "signal", "blocker") if r.get(k)]
        if meta:
            header += f" ({', '.join(meta)})"
        lines.append(header)

        if r["checked"]:
            for c in r["checked"]:
                mark = "OK" if c["passed"] else "MISMATCH"
                lines.append(
                    f"    {c['dimension']}: expected={c['expectedBand']} actual={c['actualBand']} "
                    f"({c['score']}/{c['maxScore']}) {mark}"
                )
        elif not (r["expectNoBlockers"] or r["expectBlocked"] is not None):
            lines.append("    (no dimensions named in `expected`)")

        # expectNoBlockers/expectBlocked are independent checks (both may be
        # declared, though normally only one is); report whichever fired.
        if r["expectNoBlockers"] or r["expectBlocked"] is not None:
            expectation = "blocked" if r["expectBlocked"] else "not blocked"
            actual = "blocked" if r["hardBlockers"] else "not blocked"
            mark = "OK" if r["blockersOk"] else "MISMATCH"
            lines.append(f"    blockers: expected={expectation} actual={actual} {mark}")

        # Surface the model's own reasoning whenever a blocker actually fired,
        # regardless of whether this case declared an expectation about it.
        if r["hardBlockers"]:
            for b in r["hardBlockers"]:
                lines.append(f"      - {b}")

        observed_str = " ".join(
            f"{dim}={o['score']}/{o['maxScore']}" for dim, o in r["observed"].items()
        )
        lines.append(f"    observed: {observed_str}")
        lines.append("")

    failures = [r for r in results if not r["passed"]]
    if not failures:
        lines.append("No failures.")

    return "\n".join(lines)

File: server/api/src/EvalHarness/SubscoreEval.cs
using System.Text;
using System.Text.Json;

namespace ApplicationTracker.EvalHarness;

public sealed record DimensionCheck(
    string Dimension, string ExpectedBand, string? ActualBand, int? Score, int? MaxScore)
{
    public bool Passed => ActualBand == ExpectedBand;
}

public sealed record SubscoreResult
{
    public required string Id { get; init; }
    public string? Title { get; init; }
    public List<DimensionCheck> Checked { get; init; } = [];
    public int HardBlockers { get; init; }
    public bool BlockersOk { get; init; }
    public bool Passed => BlockersOk && Checked.All(c => c.Passed);
}

/// <summary>
/// Golden-set evaluation of the Evaluator's per-dimension sub-scores, against a
/// frozen profile.
/// </summary>
/// <remarks>
/// The verdict eval asks whether the overall call is right. This asks whether
/// it is right for the right reasons — a posting can land in the correct band
/// while the dimension that should have driven it scored flat.
///
/// The profile is sent with the request and frozen in
/// <c>fixtures/golden-profile.json</c>, so the measurement does not move when
/// the real profile is edited. Candidate signal comes only from the injected
/// profile (AGENTS.md), which is what makes freezing it sufficient.
/// </remarks>
public static class SubscoreEval
{
    private static readonly string[] Dimensions =
        ["technicalFit", "engineeringExecutionFit", "sustainabilityPaceFit"];

    // Banded by percentage of each dimension's OWN maxScore, not a hardcoded
    // point value: maxScore is model-reported and dimension-specific.
    private const int LowCeiling = 28;   // < 28%   -> low
    private const int MidCeiling = 57;   // 28-57%  -> mid; above -> high

    private static readonly TimeSpan Pacing = TimeSpan.FromSeconds(3);

    /// <summary>
    /// Null when the response carried no usable score/maxScore pair — kept
    /// distinct from a real band so a missing dimension never silently reads as
    /// a pass, or as a specific band.
    /// </summary>
    public static string? BandFor(int? score, int? maxScore)
    {
        if (score is null || maxScore is null or 0) return null;
        var pct = score.Value * 100.0 / maxScore.Value;
        return pct < LowCeiling ? "low" : pct <= MidCeiling ? "mid" : "high";
    }

    public static async Task<List<SubscoreResult>> RunAsync(
        HttpClient http, List<GoldenCase> cases, string profilePath, CancellationToken ct = default)
    {
        using var profileDoc = JsonDocument.Parse(File.ReadAllText(profilePath));
        var profile = profileDoc.RootElement.Clone();

        using var rawCases = JsonDocument.Parse(File.ReadAllText(
            Path.Combine(Path.GetDirectoryName(profilePath)!, "golden-set.json")));
        var expectations = rawCases.RootElement.GetProperty("cases")
            .EnumerateArray()
            .ToDictionary(c => c.GetProperty("id").GetString()!, c => c.Clone());

        var results = new List<SubscoreResult>();

        for (var i = 0; i < cases.Count; i++)
        {
            var c = cases[i];
            if (i > 0) await Task.Delay(Pacing, ct);

            using var response = await http.PostAsJsonAsyncWithProfile(c.JdText, profile, ct);
            if (!response.IsSuccessStatusCode)
                // Aborts the whole run rather than reporting a partial result:
                // a run over fewer cases is a different denominator.
                throw new InvalidOperationException(
                    $"eval-subscore: /api/match returned {(int)response.StatusCode} for case '{c.Id}'");

            using var body = JsonDocument.Parse(await response.Content.ReadAsStringAsync(ct));
            var root = body.RootElement;

            var breakdown = root.TryGetProperty("breakdown", out var b) ? b : default;
            var blockers = root.TryGetProperty("hardBlockers", out var hb) && hb.ValueKind == JsonValueKind.Array
                ? hb.GetArrayLength() : 0;

            var spec = expectations.GetValueOrDefault(c.Id);
            var checks = new List<DimensionCheck>();

            if (spec.ValueKind == JsonValueKind.Object
                && spec.TryGetProperty("expected", out var expected)
                && expected.ValueKind == JsonValueKind.Object)
            {
                foreach (var e in expected.EnumerateObject())
                {
                    var (score, max) = ReadDimension(breakdown, e.Name);
                    checks.Add(new DimensionCheck(e.Name, e.Value.GetString() ?? "", BandFor(score, max), score, max));
                }
            }

            // expectBlocked is tri-state: true/false assert on hardBlockers
            // being non-empty/empty, and absent means this flag has no opinion.
            // Cases that test the veto itself carry no `expected` dict at all.
            var blockersOk = true;
            if (spec.ValueKind == JsonValueKind.Object)
            {
                if (spec.TryGetProperty("expectNoBlockers", out var enb)
                    && enb.ValueKind == JsonValueKind.True && blockers > 0)
                    blockersOk = false;

                if (spec.TryGetProperty("expectBlocked", out var eb)
                    && eb.ValueKind is JsonValueKind.True or JsonValueKind.False
                    && (blockers > 0) != eb.GetBoolean())
                    blockersOk = false;
            }

            results.Add(new SubscoreResult
            {
                Id = c.Id,
                Title = c.Title,
                Checked = checks,
                HardBlockers = blockers,
                BlockersOk = blockersOk,
            });

            Console.Error.WriteLine($"  [{i + 1}/{cases.Count}] {c.Id}: "
                + string.Join(" ", checks.Select(x => $"{x.Dimension}={x.ActualBand ?? "?"}")));
        }

        return results;
    }

    private static (int? Score, int? Max) ReadDimension(JsonElement breakdown, string dimension)
    {
        if (breakdown.ValueKind != JsonValueKind.Object
            || !breakdown.TryGetProperty(dimension, out var d)
            || d.ValueKind != JsonValueKind.Object)
            return (null, null);

        int? Read(string name) =>
            d.TryGetProperty(name, out var v) && v.ValueKind == JsonValueKind.Number ? v.GetInt32() : null;

        return (Read("score"), Read("maxScore"));
    }

    public static string Report(List<SubscoreResult> results)
    {
        var sb = new StringBuilder();
        sb.AppendLine();
        sb.AppendLine($"Sub-score golden set: {results.Count(r => r.Passed)}/{results.Count} passed");

        foreach (var r in results.Where(r => !r.Passed))
        {
            sb.AppendLine();
            sb.AppendLine($"  {r.Id} — {r.Title}");
            if (!r.BlockersOk)
                sb.AppendLine($"      hardBlockers: {r.HardBlockers} (not what the case expects)");
            foreach (var c in r.Checked.Where(c => !c.Passed))
                sb.AppendLine($"      {c.Dimension,-26} expected {c.ExpectedBand,-5} got "
                    + $"{c.ActualBand ?? "none",-5} ({c.Score}/{c.MaxScore})");
        }

        // Per-dimension totals: a failure concentrated in one dimension is a
        // different signal from the same count spread across all three.
        sb.AppendLine();
        sb.AppendLine("By dimension:");
        foreach (var dim in Dimensions)
        {
            var checks = results.SelectMany(r => r.Checked).Where(c => c.Dimension == dim).ToList();
            if (checks.Count > 0)
                sb.AppendLine($"    {dim,-28} {checks.Count(c => c.Passed)}/{checks.Count}");
        }

        return sb.ToString();
    }
}

internal static class HttpExtensions
{
    /// <summary>
    /// <c>POST /api/match</c> with a frozen profile override.
    /// </summary>
    /// <remarks>
    /// Hand-built rather than serialised from a typed request, because this
    /// project deliberately has no reference to Core — it measures the endpoint
    /// as a caller sees it, and a shared type would let a change to the model
    /// silently change what is being measured.
    /// </remarks>
    public static Task<HttpResponseMessage> PostAsJsonAsyncWithProfile(
        this HttpClient http, string jobDescription, JsonElement profile, CancellationToken ct)
    {
        var payload = new Dictionary<string, object?>
        {
            ["jobDescription"] = jobDescription,
            ["profile"] = profile,
        };
        return http.PostAsync("/api/match",
            new StringContent(JsonSerializer.Serialize(payload), Encoding.UTF8, "application/json"), ct);
    }
}
